subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompilationAndInstrumentation.java

```diff
@@ -17,51 +17,84 @@
 import javax.annotation.OverridingMethodsMustInvokeSuper;
 import java.io.File;
 import java.util.Map;
+import java.util.Optional;
 
 import static java.util.Objects.requireNonNull;
+import static org.gradle.internal.instrumentation.api.types.BytecodeInterceptorFilter.INSTRUMENTATION_AND_BYTECODE_REPORT;
+import static org.gradle.internal.instrumentation.api.types.BytecodeInterceptorFilter.INSTRUMENTATION_ONLY;
+import static org.gradle.internal.instrumentation.reporting.MethodInterceptionReportCollector.INTERCEPTED_METHODS_REPORT_FILE;
 
 /**
  * A base class that represents a work for compilation for Kotlin and Groovy build scripts.
  * This work unit first compiles the build script to a directory, and then instruments the directory for configuration cache and returns instrumented output.
  */
 public abstract class BuildScriptCompilationAndInstrumentation implements ImmutableUnitOfWork {
+
+    private static final CachingDisabledReason CACHING_DISABLED_FOR_PROPERTY_REPORT = new CachingDisabledReason(CachingDisabledReasonCategory.NOT_CACHEABLE, "Caching of buildscript compilation disabled due for property upgrade report");
+
+    private final ScriptSource source;
     private final ImmutableWorkspaceProvider workspaceProvider;
     private final InputFingerprinter inputFingerprinter;
     private final ClasspathElementTransformFactoryForLegacy transformFactory;
     protected final FileCollectionFactory fileCollectionFactory;
+    private final GradleCoreInstrumentationTypeRegistry gradleCoreTypeRegistry;
+    private final PropertyUpgradeReportConfig propertyUpgradeReportConfig;
 
     public BuildScriptCompilationAndInstrumentation(
+        ScriptSource source,
         ImmutableWorkspaceProvider workspaceProvider,
         FileCollectionFactory fileCollectionFactory,
         InputFingerprinter inputFingerprinter,
-        ClasspathElementTransformFactoryForLegacy transformFactory
+        ClasspathElementTransformFactoryForLegacy transformFactory,
+        GradleCoreInstrumentationTypeRegistry gradleCoreTypeRegistry,
+        PropertyUpgradeReportConfig propertyUpgradeReportConfig
     ) {
+        this.source = source;
         this.workspaceProvider = workspaceProvider;
         this.fileCollectionFactory = fileCollectionFactory;
         this.inputFingerprinter = inputFingerprinter;
         this.transformFactory = transformFactory;
+        this.gradleCoreTypeRegistry = gradleCoreTypeRegistry;
+        this.propertyUpgradeReportConfig = propertyUpgradeReportConfig;
+    }
+
+    @Override
+    public Optional<CachingDisabledReason> shouldDisableCaching(@Nullable OverlappingOutputs detectedOverlappingOutputs) {
+        // Disable caching always for property upgrade report,
+        // since there is not much use to cache report remotely, also report can contain absolute paths
+        return propertyUpgradeReportConfig.isEnabled() ? Optional.of(CACHING_DISABLED_FOR_PROPERTY_REPORT) : Optional.empty();
```

Comment:
💭 We disable cache when property upgrade report is enabled, since it contains absolute paths and I think it doesn't make sense to cache it remotely